rpc: validate fee estimation mode case insensitive

torkelrogstad opened the pull request:
The RPC docs for `send` claim that `estimate_mode` is case insensitive. This is not the case on master: 

https://github.com/bitcoin/bitcoin/blob/65c05db660b2ca1d0076b0d8573a6760b3228068/src/wallet/rpc/spend.cpp#L193-L195

This PR fixes that, by converting the estimation mode to lower case before comparing to `"unset"`. Note that the actual _parsing_ of the estimation mode already happens case insensitively, further down in the same function.

torkelrogstad committed:
rpc: validate fee estimation mode case insensitive

ismaelsadeeq reviewed (commented):
Concept ACK

Thanks for trying to solve this issue.

This patch does not fix the issue.

Behaviour on master 65c05db660b2ca1d0076b0d8573a6760b3228068
```terminal
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoind -regtest  -fallbackfee=0.0000100 -daemon
Bitcoin Core starting
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest  loadwallet abubakar-test
{
  "name": "abubakar-test"
}
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest send '{"bcrt1qledjqv5a23zjysjvk0204zgu50z2zwch9h88k8": 0.1}' 1 Economical
{
  "txid": "fdbfc6e00efef18dcfde0f0f6c428bf7dbacb776cf0f1c710d410eba20dcfcf6",
  "complete": true
}
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest send '{"bcrt1qledjqv5a23zjysjvk0204zgu50z2zwch9h88k8": 0.1}' 1           
error code: -8
error message:
Specify estimate_mode
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest send '{"bcrt1qledjqv5a23zjysjvk0204zgu50z2zwch9h88k8": 0.1}' 1 unset
error code: -8
error message:
Specify estimate_mode
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest send '{"bcrt1qledjqv5a23zjysjvk0204zgu50z2zwch9h88k8": 0.1}' 1 Unset     
{
  "txid": "937b0205df5a56414e4751b162510e17955788f09670c76086e825253ac46e23",
  "complete": true
}
```

The last RPC command on should not pass at all, because `Unset` is an invalid `estimatemode`

Same behavior on your PR head eed7296bba12dcf0df7c063d335dbf34bf358f9c

```terminal
abubakarismail@Abubakars-MacBook-Pro bitcoin % git cherry-pick eed7296bba12dcf0df7c063d335dbf34bf358f9c
Auto-merging src/wallet/rpc/spend.cpp
[01-2024-register-validation-interface-fix 3d94c6c3ca0] rpc: validate fee estimation mode case insensitive
 Author: Torkel Rogstad <torkel@rogstad.io>
 Date: Thu Jan 4 11:24:07 2024 +0100
 1 file changed, 1 insertion(+), 1 deletion(-)
abubakarismail@Abubakars-MacBook-Pro bitcoin % make                                                                                                
Making all in src
  CXX      wallet/rpc/libbitcoin_wallet_a-spend.o
  AR       libbitcoin_wallet.a
  GEN      obj/build.h
  CXX      libbitcoin_util_a-clientversion.o
  AR       libbitcoin_util.a
  CXXLD    bitcoind
  CXXLD    bitcoin-cli
  CXXLD    bitcoin-tx
  CXXLD    bitcoin-wallet
  CXXLD    bitcoin-util
  CXXLD    test/test_bitcoin
Making all in doc/man
make[1]: Nothing to be done for `all'.
make[1]: Nothing to be done for `all-am'.
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest stop                                                                  
Bitcoin Core stopping
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoind -regtest  -fallbackfee=0.0000100 -daemon                                          
Bitcoin Core starting
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest  loadwallet abubakar-test                                                
{
  "name": "abubakar-test"
}
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest send '{"bcrt1qledjqv5a23zjysjvk0204zgu50z2zwch9h88k8": 0.1}' 1 Unset
{
  "txid": "91e9f79d7b653023a522cd0bed715bc482b60eaf1036d35bb7c57d3fd630eecc",
  "complete": true
}
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest send '{"bcrt1qledjqv5a23zjysjvk0204zgu50z2zwch9h88k8": 0.1}' 1 unset
error code: -8
error message:
Specify estimate_mode
```


This patch seems to have fix the issue

```diff
diff --git a/src/wallet/rpc/spend.cpp b/src/wallet/rpc/spend.cpp
index b26dd864951..172971cd9d2 100644
--- a/src/wallet/rpc/spend.cpp
+++ b/src/wallet/rpc/spend.cpp
@@ -71,7 +71,7 @@ static void InterpretFeeEstimationInstructions(const UniValue& conf_target, cons
     } else {
         options.pushKV("fee_rate", fee_rate);
     }
-    if (!options["conf_target"].isNull() && (options["estimate_mode"].isNull() || (options["estimate_mode"].get_str() == "unset"))) {
+    if (!options["conf_target"].isNull() && (options["estimate_mode"].isNull() || (ToLower(options["estimate_mode"].get_str()) == "unset"))) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Specify estimate_mode");
     }
 }
```
```terminal
abubakarismail@Abubakars-MacBook-Pro bitcoin % make
Making all in src
  CXX      wallet/rpc/libbitcoin_wallet_a-spend.o
  AR       libbitcoin_wallet.a
  GEN      obj/build.h
  CXX      libbitcoin_util_a-clientversion.o
  AR       libbitcoin_util.a
  CXXLD    bitcoind
  CXXLD    bitcoin-cli
  CXXLD    bitcoin-tx
  CXXLD    bitcoin-wallet
  CXXLD    bitcoin-util
  CXXLD    test/test_bitcoin
Making all in doc/man
make[1]: Nothing to be done for `all'.
make[1]: Nothing to be done for `all-am'.
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest stop                                                             
Bitcoin Core stopping
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoind -regtest  -fallbackfee=0.0000100 -daemon                                  
Bitcoin Core starting
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest  loadwallet abubakar-test                                           
{
  "name": "abubakar-test"
}
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest send '{"bcrt1qledjqv5a23zjysjvk0204zgu50z2zwch9h88k8": 0.1}' 1 Unset
error code: -8
error message:
Specify estimate_mode
```


But its worth investigating why we can be able to send with an invalid `estimatemode` like `Unset`, or `uNset` in current master 65c05db660b2ca1d0076b0d8573a6760b3228068. 


Please add a functional test that catches this issue

torkelrogstad commented:
Thanks for the excellent review @ismaelsadeeq , very much appreciated. You're correct in your findings! I've updated the logic, and added functional tests for this

ismaelsadeeq commented on test/functional/wallet_send.py:
```diff
@@ -305,6 +305,33 @@ def run_test(self):
+        # verify that fee estimation modes parse case insensitively
+        self.log.info("Testing case insensitive fee estimation mode parse")
+        for mode in ["ecoNOMICAL", "economical", "ECONOMICAL"]:
+            res = self.test_send(from_wallet=w0, to_wallet=w1, amount=1,
+                estimate_mode=mode, add_to_wallet=False
+            )
+            assert res["complete"]
+
+            res = self.test_send(from_wallet=w0, to_wallet=w1, amount=1,
+                arg_estimate_mode=mode, add_to_wallet=False
+            )
+            assert res["complete"]
```
`estimatemode` is validated when passed with confirmation target
```suggestion
            res = self.test_send(from_wallet=w0, to_wallet=w1, amount=1,
                arg_conf_target=1, arg_estimate_mode=mode, add_to_wallet=False
            )
            assert_equal(res["complete"], True)
            
```

ismaelsadeeq commented on src/wallet/rpc/spend.cpp:
```diff
@@ -71,7 +71,7 @@ static void InterpretFeeEstimationInstructions(const UniValue& conf_target, cons
-    if (!options["conf_target"].isNull() && (options["estimate_mode"].isNull() || (options["estimate_mode"].get_str() == "unset"))) {
+    if (!options["conf_target"].isNull() && (options["estimate_mode"].isNull() || (ToLower(options["estimate_mode"].get_str()) == "unset"))) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Specify estimate_mode");
     }
```
I think we should also return error when `estimatemode` is passed without confirmation target and a a test for the test case ?
Maybe in its own commit or a follow-up PR.
```suggestion
    }
    if (options["conf_target"].isNull() && !options["estimate_mode"].isNull()) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, "estimate_mode should be passed with conf_target");
    }
```

ismaelsadeeq reviewed (commented):
CI failure is unrelated?

torkelrogstad committed:
rpc: validate conf_target is set alongside estimate_mode

torkelrogstad commented:
@ismaelsadeeq I fixed the tests per your comments. Note that I pass the arguments through both the `arg_` prefixed and non-prefixed versions. Figured we might as well validate that there's no behavior changes between passing as positional arguments or through `options`. 

Also added a new commits with tests that validates that `estimate_mode` is passed together with a confirmation target.

torkelrogstad commented:
I believe CI failures are unrelated. Is it possible to force CI to re-run?

ismaelsadeeq commented on src/wallet/rpc/spend.cpp:
```diff
@@ -71,9 +71,13 @@ static void InterpretFeeEstimationInstructions(const UniValue& conf_target, cons
     } else {
```
nit:
iwyu
```cpp
#include <util/strencodings.h>
```

torkelrogstad replied:
Sorry, I don't understand what you mean. Do you want me to use one of the functions in `strencodings` to manipulate the `fee_rate` parameter before sending it to `options.pushKV`?

ismaelsadeeq replied:
I meant Import what you use.
`ToLower()` function is declared here.
https://github.com/bitcoin/bitcoin/blob/4ad83ef09b772f3f1e8ea9ea3a8c43dcd9db8458/src/util/strencodings.h#L327

torkelrogstad replied:
Ah, thanks. I'm very new to C++, so I didn't realize I could even use things that weren't imported. Didn't realize I should also include a header statement, since it compiled just fine! Added the include.

ismaelsadeeq reviewed (commented):
Code review ACK 51fdb4819ebda4d4a8f64c8d36a471af65a033c9

Just a single nit, happy to re ACK when fixed.

I've tested this on regtest and it passes as expected.

ismaelsadeeq commented:
> I believe CI failures are unrelated. Is it possible to force CI to re-run?

It has to do with #29234, If you rebase on master I think the CI will be happy, but IMO there is no need since it's unrelated

torkelrogstad commented:
Rebased, let's see if that solves the CI failure.

ismaelsadeeq reviewed (approved):
Re-ACK be8ae64b82e2c5b003a8703668ce1751442288e4

laanwj commented:
Concept NACK. i think the solution here is to update the documentation. AFAIK, none of our RPC APIs have historically been case insensitive, and i don't think there's a good reason to start here. Ideally, imo, commands and field names should be considered "just blobs", not subject to unicode casing rules.

torkelrogstad commented:
Fee estimation modes are currently parsed case-insensitively here: 

https://github.com/bitcoin/bitcoin/blob/e31956980e16ad3d619022e572bdf55a4eae8716/src/util/fees.cpp#L57-L67

Which makes it possible to pass arguments case-insensitively to (at least) `estimatesmartfee`: 

```
$ bitcoin-cli estimatesmartfee 3 economical
{
  "feerate": 0.00022015,
  "blocks": 3
}

$ bitcoin-cli estimatesmartfee 3 ECONOMICAL
{
  "feerate": 0.00022015,
  "blocks": 3
}
```

Changing this would be a breaking change, and for example break client libraries like the de-facto standard Go-based `rpcclient` (from the `btcd` libraries): https://github.com/btcsuite/btcd/blob/ae5533602c46f4ea14a4da725c19cb3db7d5eb88/btcjson/walletsvrcmds.go#L118-L122

Matter of fact, this PR was actually motivated by the fact that I couldn't get the `send` RPC to properly work with `rpcclient` out-of-the-box, without resorting to tweaking the Go code. It therefore seems to me that the best solution here is to parse case-insensitively, due to other, already established practices in other RPCs

laanwj commented:
Right, there's something to be said to keep that where it already snuck in, to not break the existing interface. And i guess if the other enumeration items already are parsed case-insensitively, doing that for "UNSET" makes sense too, otherwise it's silly.

i just think the consensus is that this is not a direction the RPC interface is moving in in general, and i think introducing case-insensitive API was a mistake. But maybe getting rid of this has too much breaking impact now, I don't know.

torkelrogstad commented:
It's a fair point to not want to introduce new instances of undesirable behavior. But I **definitely** don't think we should change existing behavior. That would break ALL users that use Go for interfacing with Bitcoin Core, and maybe other languages as well

achow101 commented:
> This is probably more than you wanted to do, but I'd say it would be better to use `FeeModeFromString` to re-use the existing (case-insensitive) parsing code, than to re-implement the parsing and use hard-coded "unset" strings in all call places.

I agree. I think it would make more sense for `FeeModeFromString` to handle "unset" and to check its return value rather than having these repeated string comparisons. It's a bit odd that the estimate mode is being validated in 3 places - `InterpretFeeEstimationInstructions`, `SetFeeEstimateMode`, and `FeeModeFromString`. These should be consolidated as much as possible.

pablomartin4btc commented:
@torkelrogstad, are you still working on this?

torkelrogstad commented:
No, not really. Feel free to pick it up and get it in a mergeable state.
